### PR TITLE
Joss paper

### PR DESCRIPTION
**Pull request by martinju**
Draft for JOSS paper. 

Includes a summary of the methodology the package implements + some implementation details.

### CODE REVIEW

**Comment by nikolase90** [comment]
We should cite `Rcpp` and `data.table`.

File: inst/joss_paper/paper.Rmd [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+When using such models for real life applications, it is often crucial to understand why a certain set of features lead 
+to exactly a specific prediction.
+Most machine learning models are however so complicated and hard to understand that they are often viewed as 
+"black-boxes" producing output when provided some input.
+
+Shapley values (@Shapley53) is a concept from cooperative game theory used to fairly distribute a joint payoff among the
+cooperating players. 
+@kononenko2010efficient and later @lundberg2017unified proposed to use the Shapley value framework to explain 
+predictions by distributing the prediction value on the input features. 
+Unfortunately, established methods and implementations for explaining predictions with Shapley values like Shapley
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
Established methods and implementations for explaining predictions with Shapley values like Shapley 
```

File: inst/joss_paper/paper.Rmd [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+Shapley values (@Shapley53) is a concept from cooperative game theory used to fairly distribute a joint payoff among the
+cooperating players. 
+@kononenko2010efficient and later @lundberg2017unified proposed to use the Shapley value framework to explain 
+predictions by distributing the prediction value on the input features. 
+Unfortunately, established methods and implementations for explaining predictions with Shapley values like Shapley 
+Sampling Values (@vstrumbelj2014explaining), SHAP/Kernel SHAP (@lundberg2017unified), and to some extent TreeSHAP 
+(@lundberg2018consistent), assume that the features are independent when approximating the Shapley values for prediction
+explanation. 
+This `R`-package implements methodology proposed by @aas2019explaining to explain predictions by accounting for the
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
`shapr` implements the methodology proposed by @aas2019explaining, where predictions are explained while accounting for the 
```

File: inst/joss_paper/paper.Rmd [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+explanation. 
+This `R`-package implements methodology proposed by @aas2019explaining to explain predictions by accounting for the 
+dependence between the features, resulting in significantly more accurate approximations to the Shapley values. 
+
+
+
+# Implementation
+
+The package implements a variant of the Kernel SHAP (@lundberg2017unified) methodology for efficiently dealing with the 
+combinatorial problem related to Shapley value formula.
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
combinatorial problem related to the Shapley value formula. 
```

File: inst/joss_paper/paper.Rmd [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+The implementation supports explanation of the following models natively: `stats::lm`, `stats::glm`, `ranger::ranger`, 
+`mgcv::gam` and `xgboost::xgboost`. 
+Moreover, the package supports explanation of custom models by supplying two simple additional class functions.
+
+The user interface in the package has largely been adopted from the `R`-package `lime` (@limeRpackage). 
+The user first sets up the explainability framework for the model to explain with the `shapr` function. 
+Then the output from `shapr` is provided to the `explain` function, along with the data to explain the prediction for
+and which method should be used to estimate the aforementioned conditional expectations.
+
+The majority of the code is plain `R`, while the most time consuming parts of the code has been coded in `C++` through
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]: Cite `R`, `data.table` and `Rcpp`.

File: inst/joss_paper/paper.Rmd [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+Moreover, the package supports explanation of custom models by supplying two simple additional class functions.
+
+The user interface in the package has largely been adopted from the `R`-package `lime` (@limeRpackage). 
+The user first sets up the explainability framework for the model to explain with the `shapr` function. 
+Then the output from `shapr` is provided to the `explain` function, along with the data to explain the prediction for
+and which method should be used to estimate the aforementioned conditional expectations.
+
+The majority of the code is plain `R`, while the most time consuming parts of the code has been coded in `C++` through 
+the `Rcpp` package for computational speed up. 
+Our implementation has some minor methodological improvements to the Kernel SHAP methodology as implemented in the
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
Our implementation has methodological improvements to the Kernel SHAP methodology as implemented in the 
```

File: inst/joss_paper/paper.Rmd [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+The user first sets up the explainability framework for the model to explain with the `shapr` function. 
+Then the output from `shapr` is provided to the `explain` function, along with the data to explain the prediction for
+and which method should be used to estimate the aforementioned conditional expectations.
+
+The majority of the code is plain `R`, while the most time consuming parts of the code has been coded in `C++` through 
+the `Rcpp` package for computational speed up. 
+Our implementation has some minor methodological improvements to the Kernel SHAP methodology as implemented in the 
+`Python` package `shap` (@shapPythonpackage). 
+In addition to our package's ability to account for the feature dependence (which the `shap` package does not), 
+basic tests suggest our implementation is about 3-4 times faster.
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]: This sounds vague.

File: inst/joss_paper/paper.Rmd [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+and which method should be used to estimate the aforementioned conditional expectations.
+
+The majority of the code is plain `R`, while the most time consuming parts of the code has been coded in `C++` through 
+the `Rcpp` package for computational speed up. 
+Our implementation has some minor methodological improvements to the Kernel SHAP methodology as implemented in the 
+`Python` package `shap` (@shapPythonpackage). 
+In addition to our package's ability to account for the feature dependence (which the `shap` package does not), 
+basic tests suggest our implementation is about 3-4 times faster. 
+
+For more information about the underlying methodology of the package, we refer to the
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
For a detailed description of the underlying methodology of the package, we refer to the 
```

File: inst/joss_paper/paper.Rmd [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+The majority of the code is plain `R`, while the most time consuming parts of the code has been coded in `C++` through 
+the `Rcpp` package for computational speed up. 
+Our implementation has some minor methodological improvements to the Kernel SHAP methodology as implemented in the 
+`Python` package `shap` (@shapPythonpackage). 
+In addition to our package's ability to account for the feature dependence (which the `shap` package does not), 
+basic tests suggest our implementation is about 3-4 times faster. 
+
+For more information about the underlying methodology of the package, we refer to the 
+[paper](https://arxiv.org/abs/1903.10464) (@aas2019explaining).
+For getting started with the package, we refer to the package vignette and introductory examples available at the
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
For getting started with the package, we highly recommend the user to go through the package vignette and introductory examples available at the 
```

File: inst/joss_paper/paper.Rmd (hunk 2 above)
Comment (martinju) [reply inst/joss_paper/paper.Rmd]:
```suggestion
The `R`-package `shapr` implements the methodology proposed by @aas2019explaining, where predictions are explained while accounting for the 
```

Comment (martinju) [reply inst/joss_paper/paper.Rmd]: Ok with this @nikolase90 ?

File: inst/joss_paper/paper.Rmd (hunk 6 above)
Comment (martinju) [reply inst/joss_paper/paper.Rmd]: Indeed, it is vague. I don't think we have the space to say what our methodological improvements really are. Ok with skipping the speedup thing and keeping the rest similar to how it is? @nikolase90

File: inst/joss_paper/paper.Rmd (hunk 8 above)
Comment (martinju) [reply inst/joss_paper/paper.Rmd]:
```suggestion
For getting started with the package, we recommend the user to go through the package vignette and introductory examples available at the 
```

File: inst/joss_paper/paper.Rmd (hunk 2 above)
Comment (nikolase90) [reply inst/joss_paper/paper.Rmd]: Yep!

File: inst/joss_paper/paper.Rmd (hunk 4 above)
Comment (martinju) [reply inst/joss_paper/paper.Rmd]: Fixed in 16ff339

File: inst/joss_paper/paper.Rmd (hunk 6 above)
Comment (martinju) [reply inst/joss_paper/paper.Rmd]: Attempted to fix this in 16ff339/ d107cfa, by deleting/moving the relevant parts. What do you think @nikolase90 ?

Comment (nikolase90) [reply inst/joss_paper/paper.Rmd]: :+1:

File: inst/joss_paper/paper.Rmd [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+This allows the user to easily see the effect and importance of accounting for the feature dependence.
+
+The user interface in the package has largely been adopted from the `R`-package `lime` (@limeRpackage). 
+The user first sets up the explainability framework for the model to explain with the `shapr` function. 
+Then the output from `shapr` is provided to the `explain` function, along with the data to explain the prediction for
+and which method should be used to estimate the aforementioned conditional expectations.
+
+The majority of the code is plain `R` (@rCore), while the most time consuming operations has been coded in `C++` 
+through the `Rcpp` package (@rcppRpackage) for computational speed up. 
+For RAM efficiency and computational seed up of typical bookeeping operations in `R`, we utilize the `data.table`
```
Comment (nikolase90) [inline inst/joss_paper/paper.Rmd]:
```suggestion
For RAM efficiency and computational speed up of typical bookeeping operations in `R`, we utilize the `data.table` 
```

File: inst/joss_paper/paper.Rmd [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+for word counting in the Rmarkdwon document (250-1000 words)
+
+-->
+
+
+
+
+# Summary
+
+A common task within machine learning is to train a model which is able to predict an unknown outcome
```
Comment (aredelmeier) [inline inst/joss_paper/paper.Rmd]:
```suggestion
test
```

**Comment by martinju** [comment]
What do you think, @nikolase90, should we merge it or keep it as a separate branch while being under review?

**Comment by nikolase90** [comment]
It think we should merge it, but we'll need to do the #158 and #164 first. Is it necessary to include `apa.csl`? Or a better question, what is `apa.csl `?

**Comment by martinju** [comment]
> It think we should merge it, but we'll need to do the #158 and #164 first. Is it necessary to include `apa.csl`? Or a better question, what is `apa.csl `?

It is created by the rticles package, and needed to style the references in the paper I think. So I think we need it.

**Comment by nikolase90** [comment]
@martinju Should you remove all the comments in the files? Or do you think they will review the actual pdf?

**Comment by martinju** [comment]
> @martinju Should you remove all the comments in the files? Or do you think they will review the actual pdf?

Don't know, but anyway we don't need them anymore. Fixed in 67df871

**Pull request merged**